File: SudokuSolver - Nodejs/Solve.js
const fs = require('fs'); 

var table=[];
var puzzle=[];
var inProgress=[];
var solutions=[];
var global=0;

function Cell(x,y,num){
    this.x=x;
    this.y=y;
    this.index=this.y*9+this.x;
    this.num=num;
    this.unSolved=true;
    if(this.num!=0)
        this.unSolved=false;   

    this.checkError=function(){
        if(this.checkErrorBox()||this.checkErrorCol()||this.checkErrorRow()){

            return true;
        }
        return false;
    }
    this.checkErrorBox=function(){
        var xBox=Math.floor(Math.floor(this.index%9)/3);          //0, 1, 2 Col
        var yBox=Math.floor(Math.floor(this.index/9)/3);          //0, 1, 2 Row
        var counterBox=((yBox*3)*9)+(xBox*3);
        var invalid=[];
        for(var i=0;i<3;i++){
            
            if(invalid.includes(puzzle[counterBox])&& puzzle[counterBox] !=0){
                
                return true;
            }
            invalid.push(puzzle[counterBox]);
            counterBox++;
        }
        counterBox+=6;
        for(var i=0;i<3;i++){
            if(invalid.includes(puzzle[counterBox])&& puzzle[counterBox] !=0){
                return true;
            }
            invalid.push(puzzle[counterBox]);
            counterBox++; 
        }
        counterBox+=6;
        for(var i=0;i<3;i++){
            if(invalid.includes(puzzle[counterBox])&& puzzle[counterBox] !=0){
                return true;
            }
            invalid.push(puzzle[counterBox]);
            counterBox++;
        }
        
        
        return false;
    }
    
    this.checkErrorRow=function(){
        var iRow=Math.floor(this.index/9);
        var counter=iRow*9;
        var invalid=[];
        while(Math.floor(counter/9)==iRow){
            if(invalid.includes(puzzle[counter])&& puzzle[counter] !=0){
                return true;
            }
            invalid.push(puzzle[counter]);
            counter++;
        }
        return false;
    }
    
    this.checkErrorCol=function(){
        var iCol=Math.floor(this.index%9);
        var counter=iCol;
        var invalid=[];
        while(counter<=table.length){
            if(invalid.includes(puzzle[counter])&& puzzle[counter] !=0){
                return true;
            }
            invalid.push(puzzle[counter]);
            counter+=9;
        }
        return false;
        
    }
        
    this.increaseValue=function(){
        if(this.num==9){
            return false;
        }else{
        this.num++;
        return true;
        }
    }
    this.resetValue=function(){
        this.num=0;
    }
}



function setupBoard(){
    for(var y=0;y<9;y++){
        for(var x=0;x<9;x++){
            table[y*9+x]=new Cell(x,y,puzzle[y*9+x]);
            if(table[y*9+x].unSolved){
                inProgress.push(table[y*9+x]);
            }
        }
    }
    solutions.push(inProgress[global]);
}

function solvePuzzle(){
    var isSolved = false;
    do{
        if(solutions.length<=inProgress.length){
            var current=solutions[solutions.length-1];
            if(current.increaseValue()){
                for(var i=0;i<table.length;i++){
                    puzzle[i]=table[i].num;
                }
                if(!current.checkError()){
                    global++;
                    solutions.push(inProgress[global]);
                }
            }
            else{
                global--;
                solutions[solutions.length-1].resetValue();
                solutions.pop();
            }
        }
        else{
            isSolved = true;
        }
    }while(!isSolved)
    
}

function resetSwitch(){
    solutions=[];
    inProgress=[];
    global=0; 
}

function retrieveIndexedPuzzle(file_num)
{
    fs.readFileSync(`../Sudoku Puzzles/sudoku(${file_num}).csv`, (err, data) => { 
        if (err) throw err; 
      
        puzzle = (data.toString().split(',').map((x)=>{
            return parseInt(x);
        })); 
    }); 
}
function retrievePuzzleFromFile(file_name)
{
    let dataString = fs.readFileSync(`./${file_name}`); 
    puzzle = (dataString.toString().split(',').map((x)=>{
        return parseInt(x);
    })); 
}
function extractBoard(){
    let board=[];
    let count = 0;
    for(var index=0;index<puzzle.length;index++){
        if(puzzle[index] == 0){
            board.push((solutions[count].num));
            count+=1;
        }
        else{
            board.push((puzzle[index]));
        }
    }
    return board;
}
function checkCorrection(expected,actual){
    for(let index=0;index<actual.length;index++){
        if(expected[i] != actual[i]){
            console.log(expected);
            console.log(actual);
            return false;
        }
    }
    return true;
}
    

function runTest(filename){
    let start;
    if(typeof filename !== "undefined"){
        start = Date.now();
        retrievePuzzleFromFile(filename);
    }
    else{
        var puzzleIndex = Math.floor(Math.random()*70000);
        retrieveIndexedPuzzle(puzzleIndex);
    }
    setupBoard();
    solvePuzzle();
    let delta = Date.now() - start;
    console.log(delta/1000);
    resetSwitch();
    // let actual = grab_puzzle_solution(puzzle_index)
    // if(check_correction(expected,actual)):
    //     print(f"Puzzle {puzzle_index} was solved correctly.")
    //     print(actual)
    // else:
    //     print(f"Puzzle {puzzle_index} wasn't solved. ")
}
for(let i=0;i<10;i++){
    runTest('sudokuTest.csv');
}

File: SudokuSolver - Python/Solve.py
import random 
import math
import time
class Cell:
    def __init__(self, x,y,w,num):
        self.x=x
        self.y=y
        self.w=w
        self.index=self.y*9+self.x
        self.num=num
        self.unSolved=True
        if(self.num != 0):
            self.unSolved=False

    def checkErrorBox(self):
        global puzzle
        xBox=math.floor(math.floor(self.index%9)/3)
        yBox=math.floor(math.floor(self.index/9)/3)
        counterBox=((yBox*3)*9)+(xBox*3)
        invalid=[]
        for i in range(0,3):
            if(puzzle[counterBox] in invalid and  puzzle[counterBox] != 0):
                return True
            invalid.append(puzzle[counterBox])
            counterBox+=1
        counterBox+=6

        for i in range(0,3):
            if(puzzle[counterBox] in invalid and  puzzle[counterBox] != 0):
                return True
            invalid.append(puzzle[counterBox])
            counterBox+=1 
        counterBox+=6

        for i in range(0,3):
            if(puzzle[counterBox] in invalid and  puzzle[counterBox] != 0):
                return True
            invalid.append(puzzle[counterBox])
            counterBox+=1
        
        
        return False

    def checkErrorRow(self):
        global puzzle
        iRow=math.floor(self.index/9)
        counter=iRow*9
        invalid=[]
        while(math.floor(counter/9)==iRow):
            if(puzzle[counter] in invalid and  puzzle[counter] != 0):
                return True
            invalid.append(puzzle[counter])
            counter+=1
        return False

    def checkErrorCol(self):
        global puzzle
        global table
        iCol=math.floor(self.index%9)
        counter=iCol
        invalid=[]
        while(counter<len(table)):
            if(puzzle[counter] in invalid and  puzzle[counter] != 0):
                return True
            invalid.append(puzzle[counter])
            counter+=9
        return False

    def checkError(self):
            if(self.checkErrorBox() or self.checkErrorCol() or self.checkErrorRow()):
                return True
            return False
        
    def increaseValue(self):
        if(self.num==9):
            return False
        else:
            self.num+=1
            return True

    def resetValue(self):
        self.num=0

table = {}
puzzle = []
inProgress = []
solutions = []
global_inc = 0
def grab_puzzle_file(file_name):
    global puzzle
    f = open(file_name, "r")
    puzzle_string = f.read().split(',')
    puzzle = [ int(val) for val in puzzle_string]
    f.close()

def grab_puzzle(file_num):
    global puzzle
    f = open(f'Sudoku Puzzles/sudoku({file_num}).csv', "r")
    puzzle_string = f.read().split(',')
    puzzle = [ int(val) for val in puzzle_string]
    f.close()
    print(f"Grabbing puzzle - {file_num}.")

def grab_puzzle_solution(file_num):
    f = open(f'Sudoku Puzzles - Solutions/sudoku({file_num}).csv', "r")
    puzzle_string = f.read().split(',')
    puzzle_string = [ int(val) for val in puzzle_string]
    f.close()
    print(f"Grabbing solution to puzzle - {file_num}.")
    return puzzle_string

def set_up_grid():
    global puzzle
    global table
    global solutions
    global inProgress
    global global_inc
    print("Setting up Grid.")
    for y in range(0,9):
        for x in range(0,9):
            table[y*9+x]=Cell(x,y,0,puzzle[y*9+x])
            if(table[y*9+x].unSolved):
                inProgress.append(table[y*9+x])
    solutions.append(inProgress[global_inc])

def solve_puzzle():
    global puzzle
    global table
    global solutions
    global inProgress
    global global_inc
    isSolved = False
    while(not isSolved):
        if(len(solutions)<=len(inProgress)):
            current=solutions[len(solutions)-1]
            if(current.increaseValue()):
                for i in range(0,len(table)):
                    puzzle[i]=table[i].num
                if(not current.checkError()):
                    global_inc+=1
                    if(len(inProgress) > global_inc):
                        solutions.append(inProgress[global_inc])
                    else:
                        isSolved=True
            else:
                global_inc-=1
                solutions[len(solutions)-1].resetValue()
                solutions.pop()
        else:
            isSolved = True
                

def extract_board():
    global puzzle
    global table
    global solutions
    global inProgress
    global global_inc
    board=[]
    count = 0
    for index in range(0,len(puzzle)):
        if(puzzle[index] == 0):
            board.append((solutions[count].num))
            count+=1
        else:
            board.append((puzzle[index]))
        
    return board

def resetPuzzle():
    global puzzle
    global table
    global solutions
    global inProgress
    global global_inc
    solutions = []
    inProgress = []
    global_inc = 0
    set_up_grid()

def check_correction(expected,actual):
    for i in range(0,len(actual)):
        if(expected[i] != actual[i]):
            print(expected)
            print(actual)
            return False
    return True


def run_test():
    puzzle_index = random.randint(0, 70000) 
    grab_puzzle(puzzle_index)
    
    set_up_grid()
    solve_puzzle()
    expected = extract_board()
    actual = grab_puzzle_solution(puzzle_index)
    if(check_correction(expected,actual)):
        print(f"Puzzle {puzzle_index} was solved correctly.")
        print(actual)
    else:
        print(f"Puzzle {puzzle_index} wasn't solved. ")

def run_test_file(file_name):
    time_start = time.time()
    grab_puzzle_file(file_name)
    set_up_grid()
    solve_puzzle()
    expected = extract_board()
    # print(expected)
    print(time.time() - time_start)

for i in range(0,10):
    run_test_file('sudokuTest.csv')